lighthouse/tests/beacon_node.rs

```diff
@@ -0,0 +1,685 @@
+use beacon_node::ClientConfig as Config;
+
+use eth2_libp2p::PeerId;
+use serde_json::from_reader;
+use std::fs::File;
+use std::io::Write;
+use std::net::{IpAddr, Ipv4Addr};
+use std::path::PathBuf;
+use std::process::{Command, Output};
+use std::str::{from_utf8, FromStr};
+use std::string::ToString;
+use tempfile::TempDir;
+use types::{Checkpoint, Epoch, Hash256};
+
+const BEACON_CMD: &str = "beacon_node";
+const CONFIG_NAME: &str = "bn_dump.json";
+const DUMP_CONFIG_CMD: &str = "dump-config";
+const IMMEDIATE_SHUTDOWN_CMD: &str = "immediate-shutdown";
+
+/// Returns the `lighthouse beacon_node --immediate-shutdown` command.
+fn base_cmd() -> Command {
+    let target_dir = env!("CARGO_BIN_EXE_lighthouse");
+    let path = target_dir
+        .parse::<PathBuf>()
+        .expect("should parse CARGO_TARGET_DIR");
+
+    let mut cmd = Command::new(path);
+    cmd.arg(BEACON_CMD)
+        .arg(format!("--{}", IMMEDIATE_SHUTDOWN_CMD));
+
+    cmd
+}
+
+/// Executes a `Command`, returning a `Result` based upon the success exit code of the command.
+fn output_result(cmd: &mut Command) -> Result<Output, String> {
+    let output = cmd.output().expect("should run command");
+
+    if output.status.success() {
+        Ok(output)
+    } else {
+        Err(from_utf8(&output.stderr)
+            .expect("stderr is not utf8")
+            .to_string())
+    }
+}
+
+// Wrapper around `Command` for easier Command Line Testing.
+struct CommandLineTest {
+    cmd: Command,
+}
+impl CommandLineTest {
+    fn new() -> CommandLineTest {
+        let base_cmd = base_cmd();
+        CommandLineTest { cmd: base_cmd }
+    }
+
+    fn flag(mut self, flag: &str, value: Option<&str>) -> Self {
+        // Build the command by adding the flag and any values.
+        self.cmd.arg(format!("--{}", flag));
+        if let Some(value) = value {
+            self.cmd.arg(value);
+        }
+        self
+    }
+
+    fn run(&mut self) -> CompletedTest {
+        // Setup temp directories.
+        let tmp_dir = TempDir::new().expect("Unable to create temporary directory");
+        let tmp_path: PathBuf = tmp_dir.path().join(CONFIG_NAME);
+
+        // Add --datadir <temp_dir> --dump-config <temp_path> to cmd.
+        self.cmd
+            .arg("--datadir")
+            .arg(tmp_dir.path().as_os_str())
+            .arg(format!("--{}", DUMP_CONFIG_CMD))
+            .arg(tmp_path.as_os_str());
+
+        // Run the command.
+        let _output = output_result(&mut self.cmd).expect("Unable to run command");
```

Comment:
If we add the shutdown reason we'll need to check this `Result`.

Reply:
Could you just double-check my handling of this result please?